Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipeline.cpp

```diff
@@ -0,0 +1,4398 @@
+    fCurrentRenderPassUniforms->ambientCol = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->emissiveCol = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->specularCol = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->fogColor = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->diffuseCol = {1.0, 1.0, 1.0, 1.0};
+
+    matrix_float4x4 tXfm;
+    hsMatrix2SIMD(proj->GetTransform(), &tXfm);
+    fCurrentRenderPassUniforms->uvTransforms[0].transform = tXfm;
+    fCurrentRenderPassUniforms->uvTransforms[0].UVWSrc = proj->GetUVWSrc();
+
+    fCurrNumLayers = 1;
+    // We should have put ZNoZWrite on during export, but we didn't.
+    IHandleZMode(hsGMatState::kZNoZWrite);
+
+    // This is a bit weird - in since this isn't a material we need to build a query for the right Metal program ourselves
+    plMetalFragmentShaderDescription description;
+    memset(&description, 0, sizeof(description));
+    description.numLayers = fCurrNumLayers = 1;
+
+    description.Populate(proj, 0);
+    // DX sets the color invert when the final color should be inverted. Not sure why!
+    if (proj->GetBlendFlags() & hsGMatState::kBlendInvertFinalColor) {
+        description.blendModes[0] |= hsGMatState::kBlendInvertColor;
+    }
+
+    plMetalMaterialPassPipelineState      materialShaderState(&fDevice, vRef, description);
+    plMetalDevice::plMetalLinkedPipeline* linkedPipeline = materialShaderState.GetRenderPipelineState();
+
+    fState.fCurrentPipelineState = linkedPipeline->pipelineState;
+    fDevice.CurrentRenderCommandEncoder()->setRenderPipelineState(linkedPipeline->pipelineState);
+    fDevice.CurrentRenderCommandEncoder()->setFragmentTexture(tex->fTexture, 0);
+    MTL::SamplerState* samplerState = fDevice.SampleStateForClampFlags(hsGMatState::hsGMatClampFlags(proj->GetClampFlags()));
+    fDevice.CurrentRenderCommandEncoder()->setFragmentSamplerState(samplerState, 0);
+    fState.layerStates[0].clampFlag = hsGMatState::hsGMatClampFlags(proj->GetClampFlags());
+
+    // Okay, render it already.
+
+    render.RenderPrims();
+}
+
+// IRenderProjectionEach ///////////////////////////////////////////////////////////////////////////////////////
+// Render any lights that are to be projected onto each pass of the object.
+void plMetalPipeline::IRenderProjectionEach(const plRenderPrimFunc& render, hsGMaterial* material, int iPass, const plSpan& span, const plMetalVertexBufferRef* vRef)
+{
+    // If this is a bump map pass, forget it, we've already "done" per-pixel lighting.
+    // if( fLayerState[iPass].fMiscFlags & (hsGMatState::kMiscBumpLayer | hsGMatState::kMiscBumpChans) )
+    //    return;
+
+    // Push the LayerShadowBase override. This sets the blend
+    // to framebuffer as Add/ZNoWrite and AmbientColor = 0.
+    static plLayerLightBase layLightBase;
+
+    // For each projector:
+    int k;
+    for (k = 0; k < fProjEach.size(); k++) {
+        // Push it's projected texture as a piggyback.
```

Comment:
```suggestion
        // Push its projected texture as a piggyback.
```

"it is projected texture"